server/src/client_service.rs

```diff
@@ -63,9 +241,352 @@ impl DeqsClientApi for ClientService {
+        logger: &Logger,
+    ) -> (DeqsClientApiClient, Server, Receiver<Msg>) {
+        let server_env = Arc::new(EnvBuilder::new().build());
+        let (msg_bus_tx, msg_bus_rx) = broadcast::channel::<Msg>(1000);
+
+        let client_service =
+            ClientService::new(msg_bus_tx, order_book.clone(), logger.clone()).into_service();
+        let server_builder = ServerBuilder::new(server_env)
+            .register_service(client_service)
+            .bind_using_uri(
+                &DeqsClientUri::from_str("insecure-deqs://127.0.0.1:0").unwrap(),
+                logger.clone(),
+            );
+        let mut server = server_builder.build().expect("build server");
+        server.start();
+
+        let client_env = Arc::new(EnvBuilder::new().build());
+        let ch = ChannelBuilder::default_channel_builder(client_env).connect_to_uri(
+            &DeqsClientUri::from_str(&format!(
+                "insecure-deqs://127.0.0.1:{}",
+                server.bind_addrs().next().unwrap().1
+            ))
+            .unwrap(),
+            &logger,
+        );
+        let client_api = DeqsClientApiClient::new(ch);
+
+        (client_api, server, msg_bus_rx)
+    }
+
+    #[test_with_logger]
+    fn submit_quotes_add_orders(logger: Logger) {
+        let pair = Pair {
+            base_token_id: TokenId::from(1),
+            counter_token_id: TokenId::from(2),
+        };
+        let mut rng: StdRng = SeedableRng::from_seed([1u8; 32]);
+        let order_book = InMemoryOrderBook::default();
+        let (client_api, _server, _msg_bus_rx) =
+            create_test_client_and_server(&order_book, &logger);
+
+        let scis = (0..10)
+            .map(|_| create_sci(pair.base_token_id, pair.counter_token_id, 10, 20, &mut rng))
+            .map(|sci| mc_api::external::SignedContingentInput::from(&sci))
+            .collect::<Vec<_>>();
+
+        let req = SubmitQuotesRequest {
+            quotes: scis.clone().into(),
+            ..Default::default()
+        };
+        let resp = client_api.submit_quotes(&req).expect("submit quote failed");
+
+        assert_eq!(
+            resp.get_status_codes(),
+            vec![QuoteStatusCode::CREATED; scis.len()]
+        );
+        assert_eq!(resp.get_error_messages(), vec![""; scis.len()]);
+        let mut orders = resp
+            .get_orders()
+            .iter()
+            .map(|o| Order::try_from(o).unwrap())
+            .collect::<Vec<_>>();
+
+        let mut orders2 = order_book
+            .get_orders(&pair, .., 0)
+            .unwrap()
+            .into_iter()
+            .rev() // Orders returned in newest to oldest order
+            .collect::<Vec<_>>();
+
+        // Since orders are added in parallel, the exact order at which they get added
+        // is not determinstic.
+        orders.sort();
+        orders2.sort();
+
+        assert_eq!(orders2, orders);
+    }
+
+    #[test_with_logger]
+    fn submit_quotes_refuses_duplicate_order(logger: Logger) {
```

Comment:
It might be worth clarifying the case where 2 different requests are submitted that share a quote?

Reply:
added a test for that